File: lab spring25/python a5/taskC.py
import sys
sys.stdin = open("input.txt","r")

def getGraph(graph,m):
    u = list(map(int,input().split()))
    v = list(map(int,input().split()))

    for i in range(m):
        graph[u[i]].add(v[i])
        graph[v[i]].add(u[i])

def bfs(graph,prnt,n,s,d):
    vst = [False]*(n+1)
    vst[s] = True
    Q = [s]
    prnt[s]=s
    while bool(Q):
        at = Q.pop(0)
        for u in graph[at]:
            if not vst[u]:
                vst[u] = True
                Q.append(u)
                prnt[u] = at
                if u==d: return

def path(prnt,at):
    pth = []
    depth = -1
    while at!=prnt[at]:
        pth.append(at)
        at = prnt[at]
        depth+=1
    if at==prnt[at]:
        pth.append(at)
        depth+=1
    
    print(depth)
    while bool(pth):
        print(pth.pop(),end=" ")
    print()

def solve():
    n,m,s,d = list(map(int,input().split()))
    graph = [set() for _ in range(n+1)]
    prnt = [0]*(n+1)
    getGraph(graph,m)
    if s==d:
        print(f"0\n{s}")
        return
    bfs(graph,prnt,n,s,d)
    if prnt[d]==0:
        print(-1)
        return
    # print(graph)
    path(prnt,d)


t = 1
t = int(input())
while t!=0:
    t-=1
    solve()

File: lab spring25/assignment5/taskC.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <stack>
using namespace std;

#define rept(i,n) for(int i=0;i<n;i++)
#define vi vector<int>
#define vb vector<bool>
#define pqi priority_queue<int>
#define br cout << '\n'

void makeGraph(pqi *graph, int m){
    vi u(m),v(m);
    rept(i,m) cin >> u[i];
    rept(i,m) cin >> v[i];
    rept(i,m) {
        graph[u[i]].push(-v[i]);
        graph[v[i]].push(-u[i]);
    }
}

void bfs(pqi *graph,vi &prent,int n, int s, int d){
    vb vst(n+1,false); vst[s] = true;
    queue<int> Q; Q.push(s);

    while (!Q.empty())
    {
        int at = Q.front(); Q.pop();
        pqi &g = graph[at];

        while(!g.empty()) {
           int u = -g.top(); g.pop();
            if(!vst[u]){
                vst[u] = true;
                Q.push(u);
                prent[u] = at;
            }
        }
    }
}

void path(vi &prent,int at){
    stack<int> pth;
    int dpth = -1;
    while (at!=prent[at]){
        pth.push(at);
        dpth+=(at!=0);
        at = prent[at];
    }
    cout << dpth << '\n';

    while (!pth.empty()){
        cout << pth.top() << ' ';
        pth.pop();
    }
    br;
}

void solve(){
    int n,m,s,d;
    cin >> n >> m >> s >> d;
    pqi graph[n+1];
    makeGraph(graph,m);
    vi prent(n+1);
    if(s==d){
        cout <<"0\n"<< s << '\n';
        return;
    }
    bfs(graph,prent,n,s,d);
    if(prent[d]==0){
        cout << "-1\n";
        return;
    }
    path(prent,d);
}

int main(){
    ios::sync_with_stdio(false);cin.tie(nullptr);cout.tie(nullptr);
    freopen("input.txt","r",stdin);
    int t{1};
    cin >> t;
    while(t--) solve();
}